File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/model/inheritance/RoleBase.java
package org.babyfish.jimmer.sql.model.inheritance;

import org.babyfish.jimmer.sql.*;
import org.babyfish.jimmer.sql.model.calc.RolePermissionCountResolver;

import java.util.List;

@MappedSuperclass
public interface RoleBase extends NamedEntity {

    @OneToMany(mappedBy = "role")
    List<Permission> getPermissions();

    @ManyToMany(mappedBy = "roles")
    List<Administrator> administrators();

    @Transient(RolePermissionCountResolver.class)
    int getPermissionCount();

    @IdView("permissions")
    List<Long> permissionIds();
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/inheritance/RoleBase.kt
package org.babyfish.jimmer.sql.kt.model.inheritance

import org.babyfish.jimmer.sql.ManyToMany
import org.babyfish.jimmer.sql.MappedSuperclass
import org.babyfish.jimmer.sql.OneToMany

@MappedSuperclass
interface RoleBase : NamedEntity {

    @OneToMany(mappedBy = "role")
    val permissions: List<Permission>

    @ManyToMany(mappedBy = "roles")
    val administrators: List<Administrator>
}